core/src/test/java/org/apache/iceberg/variants/TestSerializedPrimitives.java

```diff
@@ -0,0 +1,465 @@
+
+    assertThat(value.type()).isEqualTo(PhysicalType.INT16);
+    assertThat(value.get()).isEqualTo((short) -1);
+  }
+
+  @Test
+  public void testInt32() {
+    VariantPrimitive<?> value =
+        SerializedPrimitive.from(
+            new byte[] {primitiveHeader(5), (byte) 0xD2, 0x02, (byte) 0x96, 0x49});
+
+    assertThat(value.type()).isEqualTo(PhysicalType.INT32);
+    assertThat(value.get()).isEqualTo(1234567890);
+  }
+
+  @Test
+  public void testNegativeInt32() {
+    VariantPrimitive<?> value =
+        SerializedPrimitive.from(
+            new byte[] {primitiveHeader(5), (byte) 0xFF, (byte) 0xFF, (byte) 0xFF, (byte) 0xFF});
+
+    assertThat(value.type()).isEqualTo(PhysicalType.INT32);
+    assertThat(value.get()).isEqualTo(-1);
+  }
+
+  @Test
+  public void testInt64() {
+    VariantPrimitive<?> value =
+        SerializedPrimitive.from(
+            new byte[] {
+              primitiveHeader(6),
+              (byte) 0xB1,
+              0x1C,
+              0x6C,
+              (byte) 0xB1,
+              (byte) 0xF4,
+              0x10,
+              0x22,
+              0x11
+            });
+
+    assertThat(value.type()).isEqualTo(PhysicalType.INT64);
+    assertThat(value.get()).isEqualTo(1234567890987654321L);
+  }
+
+  @Test
+  public void testNegativeInt64() {
+    VariantPrimitive<?> value =
+        SerializedPrimitive.from(
+            new byte[] {
+              primitiveHeader(6),
+              (byte) 0xFF,
+              (byte) 0xFF,
+              (byte) 0xFF,
+              (byte) 0xFF,
+              (byte) 0xFF,
+              (byte) 0xFF,
+              (byte) 0xFF,
+              (byte) 0xFF
+            });
+
+    assertThat(value.type()).isEqualTo(PhysicalType.INT64);
+    assertThat(value.get()).isEqualTo(-1L);
+  }
+
+  @Test
+  public void testDouble() {
+    VariantPrimitive<?> value =
+        SerializedPrimitive.from(
+            new byte[] {
+              primitiveHeader(7),
+              (byte) 0xB1,
+              0x1C,
+              0x6C,
+              (byte) 0xB1,
+              (byte) 0xF4,
+              0x10,
+              0x22,
+              0x11
+            });
+
+    assertThat(value.type()).isEqualTo(PhysicalType.DOUBLE);
+    assertThat(value.get()).isEqualTo(Double.longBitsToDouble(1234567890987654321L));
+  }
+
+  @Test
+  public void testNegativeDouble() {
+    VariantPrimitive<?> value =
+        SerializedPrimitive.from(
+            new byte[] {primitiveHeader(7), 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, (byte) 0x80});
+
+    assertThat(value.type()).isEqualTo(PhysicalType.DOUBLE);
+    assertThat(value.get()).isEqualTo(-0.0D);
+  }
+
+  @Test
+  public void testDecimal4() {
+    VariantPrimitive<?> value =
+        SerializedPrimitive.from(
+            new byte[] {primitiveHeader(8), 0x04, (byte) 0xD2, 0x02, (byte) 0x96, 0x49});
+
+    assertThat(value.type()).isEqualTo(PhysicalType.DECIMAL4);
+    assertThat(value.get()).isEqualTo(new BigDecimal("123456.7890"));
```

Comment:
Based on the spec, DECIMAL4 should have the number with the precision between 1 and 9. 123456.7890 with precision 10 should store in DECIMAL8.